File: scanner-python/app/services/virustotal.py
import httpx
import base64
from app.config import settings
from app.utils.normalize import normalize_virustotal, utc_now
from app.models.shared import SourceMetadata

VIRUSTOTAL_IP_URL = "https://www.virustotal.com/api/v3/ip_addresses"
VIRUSTOTAL_DOMAIN_URL = "https://www.virustotal.com/api/v3/domains"
VIRUSTOTAL_URL_URL = "https://www.virustotal.com/api/v3/urls"
VIRUSTOTAL_FILE_URL = "https://www.virustotal.com/api/v3/files"

async def fetch_virustotal(ip: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="virustotal",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{VIRUSTOTAL_IP_URL}/{ip}",
                headers={
                    "x-apikey": settings.virustotal_api_key,
                    "Accept": "application/json"
                },
                timeout=10.0
            )
            response.raise_for_status()
            raw = response.json()
            normalized = normalize_virustotal(raw)
            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source

async def fetch_virustotal_domain(domain: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="virustotal",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{VIRUSTOTAL_DOMAIN_URL}/{domain}",
                headers={
                    "x-apikey": settings.virustotal_api_key,
                    "Accept": "application/json"
                },
                timeout=10.0
            )
            response.raise_for_status()
            raw = response.json()

            stats = raw.get("data", {}).get("attributes", {}).get("last_analysis_stats", {})
            categories = raw.get("data", {}).get("attributes", {}).get("categories", {})

            normalized = {
                "malicious_votes": stats.get("malicious"),
                "harmless_votes": stats.get("harmless"),
                "suspicious_votes": stats.get("suspicious"),
                "last_analysis_date": raw.get("data", {}).get("attributes", {}).get("last_analysis_date"),
                "categories": list(categories.values()) if categories else [],
            }

            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source

async def fetch_virustotal_url(url: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="virustotal",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        url_id = base64.urlsafe_b64encode(url.encode()).decode().rstrip("=")

        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{VIRUSTOTAL_URL_URL}/{url_id}",
                headers={
                    "x-apikey": settings.virustotal_api_key,
                    "Accept": "application/json"
                },
                timeout=10.0
            )
            response.raise_for_status()
            raw = response.json()

            stats = raw.get("data", {}).get("attributes", {}).get("last_analysis_stats", {})

            normalized = {
                "malicious_votes": stats.get("malicious"),
                "harmless_votes": stats.get("harmless"),
                "suspicious_votes": stats.get("suspicious"),
                "last_analysis_date": raw.get("data", {}).get("attributes", {}).get("last_analysis_date"),
            }

            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source

async def fetch_virustotal_hash(hash_value: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="virustotal",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{VIRUSTOTAL_FILE_URL}/{hash_value}",
                headers={
                    "x-apikey": settings.virustotal_api_key,
                    "Accept": "application/json"
                },
                timeout=10.0
            )
            response.raise_for_status()
            raw = response.json()

            attrs = raw.get("data", {}).get("attributes", {})
            stats = attrs.get("last_analysis_stats", {})
            threat = attrs.get("popular_threat_classification", {})

            normalized = {
                "malicious_votes": stats.get("malicious"),
                "harmless_votes": stats.get("harmless"),
                "suspicious_votes": stats.get("suspicious"),
                "last_analysis_date": attrs.get("last_analysis_date"),
                "file_type": attrs.get("type_description"),
                "file_size": attrs.get("size"),
                "meaningful_name": attrs.get("meaningful_name"),
                "threat_label": threat.get("suggested_threat_label"),
            }

            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source

File: scanner-go/internal/services/virustotal.go
package services

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/Ahlyx/scanner-go/internal/models"
)

const (
	vtIPURL     = "https://www.virustotal.com/api/v3/ip_addresses"
	vtDomainURL = "https://www.virustotal.com/api/v3/domains"
	vtURLURL    = "https://www.virustotal.com/api/v3/urls"
	vtFileURL   = "https://www.virustotal.com/api/v3/files"
)

func vtGet(apiKey, url string) ([]byte, error, int) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err, 0
	}
	req.Header.Set("x-apikey", apiKey)
	req.Header.Set("Accept", "application/json")

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err, 0
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	return body, err, resp.StatusCode
}

func FetchVirusTotalIP(apiKey, ip string) (*models.VTIPData, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "virustotal", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	body, err, status := vtGet(apiKey, fmt.Sprintf("%s/%s", vtIPURL, ip))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	if status != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", status))
		return nil, meta
	}

	var raw struct {
		Data struct {
			Attributes struct {
				LastAnalysisStats struct {
					Malicious  *int `json:"malicious"`
					Harmless   *int `json:"harmless"`
					Suspicious *int `json:"suspicious"`
				} `json:"last_analysis_stats"`
				LastAnalysisDate         *int64 `json:"last_analysis_date"`
				PopularThreatClassification struct {
					SuggestedThreatLabel *string `json:"suggested_threat_label"`
				} `json:"popular_threat_classification"`
			} `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	attrs := raw.Data.Attributes
	malware := []string{}
	if attrs.PopularThreatClassification.SuggestedThreatLabel != nil {
		malware = []string{*attrs.PopularThreatClassification.SuggestedThreatLabel}
	}

	meta.Success = true
	return &models.VTIPData{
		MaliciousVotes:    attrs.LastAnalysisStats.Malicious,
		HarmlessVotes:     attrs.LastAnalysisStats.Harmless,
		SuspiciousVotes:   attrs.LastAnalysisStats.Suspicious,
		LastAnalysisDate:  attrs.LastAnalysisDate,
		AssociatedMalware: malware,
	}, meta
}

func FetchVirusTotalDomain(apiKey, domain string) (*models.DomainVTData, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "virustotal", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	body, err, status := vtGet(apiKey, fmt.Sprintf("%s/%s", vtDomainURL, domain))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	if status != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", status))
		return nil, meta
	}

	var raw struct {
		Data struct {
			Attributes struct {
				LastAnalysisStats struct {
					Malicious  *int `json:"malicious"`
					Harmless   *int `json:"harmless"`
					Suspicious *int `json:"suspicious"`
				} `json:"last_analysis_stats"`
				LastAnalysisDate *int64            `json:"last_analysis_date"`
				Categories       map[string]string `json:"categories"`
			} `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	cats := []string{}
	for _, v := range raw.Data.Attributes.Categories {
		cats = append(cats, v)
	}

	meta.Success = true
	return &models.DomainVTData{
		MaliciousVotes:   raw.Data.Attributes.LastAnalysisStats.Malicious,
		HarmlessVotes:    raw.Data.Attributes.LastAnalysisStats.Harmless,
		SuspiciousVotes:  raw.Data.Attributes.LastAnalysisStats.Suspicious,
		LastAnalysisDate: raw.Data.Attributes.LastAnalysisDate,
		Categories:       cats,
	}, meta
}

func FetchVirusTotalURL(apiKey, targetURL string) (*models.URLVTData, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "virustotal", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	// VT URL lookup: base64url-encode the URL (no padding), same as Python.
	encoded := base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString([]byte(targetURL))
	encoded = strings.TrimRight(encoded, "=")

	body, err, status := vtGet(apiKey, fmt.Sprintf("%s/%s", vtURLURL, encoded))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	if status != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", status))
		return nil, meta
	}

	var raw struct {
		Data struct {
			Attributes struct {
				LastAnalysisStats struct {
					Malicious  *int `json:"malicious"`
					Harmless   *int `json:"harmless"`
					Suspicious *int `json:"suspicious"`
				} `json:"last_analysis_stats"`
				LastAnalysisDate *int64 `json:"last_analysis_date"`
			} `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	meta.Success = true
	return &models.URLVTData{
		MaliciousVotes:   raw.Data.Attributes.LastAnalysisStats.Malicious,
		HarmlessVotes:    raw.Data.Attributes.LastAnalysisStats.Harmless,
		SuspiciousVotes:  raw.Data.Attributes.LastAnalysisStats.Suspicious,
		LastAnalysisDate: raw.Data.Attributes.LastAnalysisDate,
	}, meta
}

func FetchVirusTotalHash(apiKey, hash string) (*models.HashVTData, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "virustotal", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	body, err, status := vtGet(apiKey, fmt.Sprintf("%s/%s", vtFileURL, hash))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	if status != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", status))
		return nil, meta
	}

	var raw struct {
		Data struct {
			Attributes struct {
				LastAnalysisStats struct {
					Malicious  *int `json:"malicious"`
					Harmless   *int `json:"harmless"`
					Suspicious *int `json:"suspicious"`
				} `json:"last_analysis_stats"`
				LastAnalysisDate            *int64  `json:"last_analysis_date"`
				TypeDescription             *string `json:"type_description"`
				Size                        *int64  `json:"size"`
				MeaningfulName              *string `json:"meaningful_name"`
				PopularThreatClassification struct {
					SuggestedThreatLabel *string `json:"suggested_threat_label"`
				} `json:"popular_threat_classification"`
			} `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	attrs := raw.Data.Attributes
	meta.Success = true
	return &models.HashVTData{
		MaliciousVotes:   attrs.LastAnalysisStats.Malicious,
		HarmlessVotes:    attrs.LastAnalysisStats.Harmless,
		SuspiciousVotes:  attrs.LastAnalysisStats.Suspicious,
		LastAnalysisDate: attrs.LastAnalysisDate,
		FileType:         attrs.TypeDescription,
		FileSize:         attrs.Size,
		MeaningfulName:   attrs.MeaningfulName,
		ThreatLabel:      attrs.PopularThreatClassification.SuggestedThreatLabel,
	}, meta
}
